ktor-server/ktor-server-core/jvm/src/io/ktor/application/newapi/KtorPlugin.kt

```diff
@@ -0,0 +1,655 @@
+     * while for logging and monitoring it is important to observe the pure (untouched) data.
+     * */
+    public fun monitoring(block: suspend CallExecution.(ApplicationCall) -> Unit): Unit
+
+    /**
+     * Defines what to do with the call in case request was not handled for some reason.
+     * Usually, it is handy to use fallback { call -> ...} to set the response with some message or status code,
+     * or to throw an exception.
+     * */
+    public fun fallback(block: suspend CallExecution.(ApplicationCall) -> Unit): Unit
+}
+
+public interface OnReceive {
+    /**
+     * Define how current [KtorPlugin] should transform data received from a client.
+     * */
+    public operator fun invoke(block: suspend ReceiveExecution.(ApplicationCall) -> Unit): Unit
+
+    /**
+     * Defines actions to perform before any transformations were made to the received content. It is useful for caching.
+     * */
+    public fun before(block: suspend ReceiveExecution.(ApplicationCall) -> Unit): Unit
+}
+
+public interface OnResponse {
+    /**
+     * Do transformations of the data. Example: you can write a custom serializer using this method.
+     * (Note: it is handy to also use [Execution.proceedWith] for this scenario)
+     * */
+    public operator fun invoke(block: suspend ResponseExecution.(ApplicationCall) -> Unit): Unit
+
+    /**
+     * Allows to use the direct result of call processing (see [OnCall.invoke]) and prepare data before sending a response will be executed.
+     * */
+    public fun before(block: suspend ResponseExecution.(ApplicationCall) -> Unit): Unit
+
+    /**
+     * Allows to calculate some statistics on the data that was already sent to a client, or to handle errors.
+     * (See [Metrics], [CachingHeaders], [StatusPages] features as examples).
+     * */
+    public fun after(block: suspend ResponseExecution.(ApplicationCall) -> Unit): Unit
+}
+
+public interface PluginContext {
+    public val onCall: OnCall
+    public val onReceive: OnReceive
+    public val onResponse: OnResponse
+}
+
+/**
+ * Compatibility class. It describes how (with what new functionality) some particular phase should be intercepted.
+ * It is a wrapper over pipeline.intercept(phase) { ... } and is needed to hide old Plugins API functionality
+ * */
+public data class Interception<T : Any>(
+    val phase: PipelinePhase,
+    val action: (Pipeline<T, ApplicationCall>) -> Unit
+)
+
+/**
+ * Compatibility class. Interception class for Call phase
+ * */
+public typealias CallInterception = Interception<Unit>
+
+/**
+ * Compatibility class. Interception class for Receive phase
+ * */
+
+public typealias ReceiveInterception = Interception<ApplicationReceiveRequest>
+
+/**
+ * Compatibility class. Interception class for Send phase
+ * */
+public typealias ResponseInterception = Interception<Any>
+
+/**
+ * Compatibility class. Every plugin that needs to be compatible with `beforePlugin(...)` and `afterPlugin(...)`
```

Comment:
Do we need this API at all? How hard is it to migrate old features to the new API? Is it really harder than implementing this compatibility API? 
If you think that it is needed, please provide a test for this API.

Reply:
Yes, we need